Eventually connect to NODE_NETWORK_LIMITED peers

jonasschnelli opened the pull request:
Eventually connect to peers signalling NODE_NETWORK_LIMITED if we are out of IBD.
Accept and relay NODE_NETWORK_LIMITED peers in addrman.

jonasschnelli committed:
[QA] Allow addrman loopback tests (add debug option -addrmantest)

jonasschnelli commented:
Rebased, overhauled and removed WIP tag.
Conforms now to BIP159.
Added 0.16 milestone tag.

jonasschnelli commented:
Added a fingerprinting counter-measure which protects from leaking the prune depth by finding it through getdata requests (21c347c8c1a9ee4803b8f8f124185c218e01db76).

fanquake commented on src/net_processing.cpp:
```diff
@@ -1006,6 +1020,13 @@ void static ProcessGetData(CNode* pfrom, const Consensus::Params& consensusParam
                     pfrom->fDisconnect = true;
                     send = false;
                 }
+                // Avoid leaking prune-height by never send blocks below the NODE_NETWORK_LIMITED threshold
```
nit: sending

fanquake commented on src/validation.h:
```diff
@@ -203,6 +203,8 @@ extern bool fPruneMode;
 extern uint64_t nPruneTarget;
 /** Block files containing a block-height within MIN_BLOCKS_TO_KEEP of chainActive.Tip() will not be pruned. */
 static const unsigned int MIN_BLOCKS_TO_KEEP = 288;
+/** Minimum blocks to must be available to signal NODE_NETWORK_LIMITED */
```
s/to must be available/required

jonasschnelli commented:
Fixed travis trailing whitespace in test issue.

TheBlueMatt commented on src/net_processing.h:
```diff
@@ -61,4 +61,17 @@ bool GetNodeStateStats(NodeId nodeid, CNodeStateStats &stats);
+ * @param[in]   interrupt       Interrupt condition for processing threads
+ * @return                      True if there is more work to be done
+ */
+bool SendMessages(CNode* pto, CConnman& connman, const std::atomic<bool>& interrupt);
```
Looks like a rebase error - these functions were moved to the PeerLogicValidation.

TheBlueMatt commented on src/net_processing.cpp:
```diff
@@ -3305,6 +3314,43 @@ bool PeerLogicValidation::SendMessages(CNode* pto, std::atomic<bool>& interruptM
+        nLocalServices = ServiceFlags(nLocalServices & ~NODE_NETWORK);
+
+        // don't signal NODE_NETWORK_LIMITED during initial block download
+        if (IsInitialBlockDownload()) {
```
Hmm, this feels pretty strange to me. Service bits indicate a commitment to provide certain services, not your current state. More importantly, IBD is always neccessarily fuzzy - we can't know how "in sync" we are.

TheBlueMatt commented on src/net_processing.cpp:
```diff
@@ -3305,6 +3314,43 @@ bool PeerLogicValidation::SendMessages(CNode* pto, std::atomic<bool>& interruptM
+        // currently, the prune mode can't be changed during runtime
+        // implement the service bit switch for possible future flexibility
+        nLocalServices = ServiceFlags(nLocalServices | NODE_NETWORK);
+        nLocalServices = ServiceFlags(nLocalServices & ~NODE_NETWORK_LIMITED);
```
Why? The BIP seems somewhat unclear here, but I'd stronlgy prefer that NODE_NETWORK_LIMITED simply be always set to indicate that we can always provide the most recent 288 blocks, irrespective of if we're pruned or not.

TheBlueMatt commented on src/net_processing.cpp:
```diff
@@ -3305,6 +3314,43 @@ bool PeerLogicValidation::SendMessages(CNode* pto, std::atomic<bool>& interruptM
+        }
+
+        // check how many blocks we have available (backwards from tip)
+        // don't go futher down then chainActive.Height()-NODE_NETWORK_LIMITED
```
This seems entirely redundant - even if by super rare chance we end up due to reorg short of 288 blocks, we won't be *very* short of 288 blocks, and will be caught back up within the next block or two, might as well just always set NODE_NETWORK_LIMITED (you can static_assert MIN_BLOCKS_TO_KEEP is at least 288 if you really want to have a check).

TheBlueMatt commented on src/protocol.h:
```diff
@@ -250,8 +250,8 @@ enum ServiceFlags : uint64_t {
-    // set by all Bitcoin Core nodes, and is unset by SPV clients or other peers that just want
-    // network services but don't provide them.
+    // set by all Bitcoin Core non pruned nodes, and is unset by SPV clients or other peers
+    // that just want network services but don't provide them.
```
If you're gonna update this maybe overhaul the text? The concept of "want network services but don't provide them" doesn't really make sense anymore (especially with NODE_NETWORK_LIMITED). Maybe just say "indicates peer is capable of serving the full historical block chain".

TheBlueMatt commented on src/net.cpp:
```diff
@@ -2672,6 +2672,11 @@ ServiceFlags CConnman::GetLocalServices() const
     return nLocalServices;
 }
 
+void CConnman::SetLocalServices(ServiceFlags flagsIn)
```
I'm entirely unconvinced NODE_NETWORK_LIMITED needs this (and would prefer to avoid adding it until we need it, though its no big deal to add it).

TheBlueMatt commented on src/rpc/blockchain.cpp:
```diff
@@ -895,6 +897,13 @@ UniValue pruneblockchain(const JSONRPCRequest& request)
+    // check if we need to change the local service flags in order to signal NODE_NETWORK_LIMITED flags
+    if(g_connman) {
+        ServiceFlags nFlags = g_connman->GetLocalServices();
+        CheckLocalServices(nFlags);
```
Why do we need to do this? You can't prune deeper than 128 and NODE_NETWORK is already unset for manually-pruned nodes.

TheBlueMatt commented on src/net_processing.cpp:
```diff
@@ -1020,6 +1020,13 @@ void static ProcessGetData(CNode* pfrom, const Consensus::Params& consensusParam
                 }
+                // Avoid leaking prune-height by never send blocks below the NODE_NETWORK_LIMITED threshold
+                if (send && !pfrom->fWhitelisted && (
+                        (((connman->GetLocalServices() & NODE_NETWORK_LIMITED) == NODE_NETWORK_LIMITED) && (chainActive.Tip()->nHeight - mi->second->nHeight > (int)NODE_NETWORK_LIMITED_MIN_BLOCKS) )
```
This should use pfrom->GetLocalServices() not connman->GetLocalServices() - our decisions should be based on what we told the peer we support, not what we would tell a new peer to support.

TheBlueMatt reviewed (commented):
I think this PR is significantly needlessly over-complicated (mostly we dont allow nodes to go below 288 blocks, no need to check for it in so many places and update our services accordingly). Excited to see this go in, though.

jonasschnelli commented:
Thanks for the review @TheBlueMatt. I agree with all you points and most things can now go away since the NODE_NETWORK_LIMITED_HIGH flag (1152 blocks minimum) is gone.

Will overhaul and simplify it now.

jonasschnelli commented:
Simplified the PR and addressed @TheBlueMatt and @fanquake points.

TheBlueMatt commented on src/init.cpp:
```diff
@@ -1590,6 +1590,8 @@ bool AppInitMain(boost::thread_group& threadGroup, CScheduler& scheduler)
         LogPrintf("Unsetting NODE_NETWORK on prune mode\n");
         nLocalServices = ServiceFlags(nLocalServices & ~NODE_NETWORK);
+        // always set NODE_NETWORK_LIMITED (BIP159) in pruned mode
+        nLocalServices = ServiceFlags(nLocalServices | NODE_NETWORK_LIMITED);
```
Hmm, I thought we were just going to alays set this? Why bother *only* setting it when pruned - a node that is not pruned is also capable of serving the last 288 blocks, which is the way I read the definition of NODE_NETWORK_LIMITED.

theuni replied:
I agree. Came to complain about the same thing.

NODE_NETWORK implies NODE_NETWORK_LIMITED, so _LIMITED should always be set.
Otherwise, we're introducing negative service flags, which is a bad idea imo.

jonasschnelli replied:
Okay. Agree... will change it.

theuni commented on src/net_processing.cpp:
```diff
@@ -1302,7 +1312,7 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
         }
         pfrom->nStartingHeight = nStartingHeight;
-        pfrom->fClient = !(nServices & NODE_NETWORK);
+        pfrom->fClient = (!(nServices & NODE_NETWORK) && !(nServices & NODE_NETWORK_LIMITED));
```
I'd rather not use fClient as a proxy for NODE_NETWORK_LIMITED. It deserves its own classification.

theuni replied:
Yes, this needs its own bool so that fClient gets checked when testing for our initial headers-sync peer. Then, fPreferredDownload is only set to false if we're still in IBD.

That way, we request from this peer as if they were a regular full node as long as we're synced.

jonasschnelli replied:
IMO we want pruned peers for header sync... but right, we need a protection to not request blocks from limited peers during IBD.
What about adding an `fLimitedNode` bool and check it at the point we populate vGetData with block requests?

Seems also simpler and more constant (NODE_NETWORK_LIMITED is not really a "client").

I changed this PR and applied my approach which I have written above.

theuni commented on test/functional/test_framework/mininode.py:
```diff
@@ -221,19 +221,24 @@ def ToHex(obj):
         self.port = 0
 
-    def deserialize(self, f):
+    def deserialize(self, f, without_time=False):
```
nit: a default of without_foo = False isn't no bad thing for nobody.

with_time=True please :)

jimpo commented on src/net.cpp:
```diff
@@ -1819,13 +1823,19 @@ void CConnman::ThreadOpenConnections(const std::vector<std::string> connect)
+
             // only consider nodes missing relevant services after 40 failed attempts and only if less than half the outbound are up.
             ServiceFlags nRequiredServices = nRelevantServices;
             if (nTries >= 40 && nOutbound < (nMaxOutbound >> 1)) {
```
In this clause, should `nRequiredServices` get set to NODE_NETWORK_LIMITED if peer is in sync. Essentially, replace `REQUIRED_SERVICES` below with a value that goes from NODE_NETWORK to NODE_NETWORK_LIMITED when peer is out of IBD?

jimpo reviewed (commented):
This change prefers connecting to full nodes rather than pruned nodes even if fully synced. Why is that the desired approach?

I find that the required services/relevant services/alternative services stuff is getting confusing. If I'm reading it right, the order of preference is:

- NODE_NETWORK & NODE_WITNESS
- NODE_NETWORK_LIMITED & NODE_WITNESS (`if !IsInitialDownload`)
- NODE_NETWORK

It seems like there should be a way to simplify this, where the ConnManager tracks the ranked preferences of peer services with the nOutbound quota and nTries threshold for each.

jonasschnelli commented:
> This change prefers connecting to full nodes rather than pruned nodes even if fully synced. Why is that the desired approach?
> [...]
> I find that the required services/relevant services/alternative services stuff is getting confusing. If I'm reading it right, the order of preference is... [...]

We probably should avoid an abrupt change for pruned peers. This approach (alternative services) would result is a slow usage of the current deployed (and then upgraded) pruned peers. Otherwise, pruned peers may see a significant overall usage (bandwidth / CPU / disk-usage) bump.

Since there is no toggle to disable NODE_NETWORK_LIMITED signalling in pruned mode, I think this (slow) approach would be wise.

jonasschnelli commented on src/net_processing.cpp:
```diff
@@ -818,6 +818,16 @@ void PeerLogicValidation::UpdatedBlockTip(const CBlockIndex *pindexNew, const CB
+
+    // update service flags
+    {
+        LOCK(cs_main);
```
I think we should not connect to NODE_NETWORK_LIMITED peers if IBD is still ongoing. NODE_NETWORK_LIMITED implementations really should make sure to not serve deeper then >288 blocks (fingerprinted). Seems useless to me connect until clear IBD.

jimpo replied:
Yeah, I understand that, but unless I'm missing something, this code is run every time there is a new block after IBD is done instead of happening only once. It seems possible to change the logic to exit before obtaining the lock if AlternativeServices has already been switched from NODE_NETWORK to NODE_NETWORK_LIMITED.

jonasschnelli commented:
Addressed @theuni & @TheBlueMatt points. Now mentions BIP159 in `protocol.h` (found by @jimpo).

theuni commented:
"Set NODE_NETWORK_LIMITED bit in pruned mode" commit message is confusing now. Mind rewording it to reflect the new behavior?

jonasschnelli commented:
@theuni: overhauled that commit message.

jimpo commented on src/net.cpp:
```diff
@@ -1819,13 +1823,19 @@ void CConnman::ThreadOpenConnections(const std::vector<std::string> connect)
                 continue;
 
+            // only consider nodes matching alternative and missing relevant services after 10 failed attempts and only if less than a fourth of the outbound are up.
+            if ( ((addr.nServices & nAlternativeServices) == nAlternativeServices) && (addr.nServices & nRelevantServices) != nRelevantServices && (nTries < 20 || nOutbound >= (nMaxOutbound >> 2)))
```
Seems that we would never connect to limited nodes because of the code on lines 1840-1847. `nRequiredServices` is set to either `nRelevantServices` or `REQUIRED_SERVICES`, never `nAlternativeServices`.

TheBlueMatt commented:
The nRelevantServices stuff is somewhat already a mess, and trying to work within the current code is just getting messier :/. I spent a bit of time rewriting it and (mostly) like the outcome, its at https://github.com/TheBlueMatt/bitcoin/commit/2017-09-service-flags-cleanups - and I think basing this PR on it would simplify this one a lot. It sucks to add some functions to protocol.h/cpp, especially if they have knowledge of IBD state, but that should be able to move soon once @theuni makes net.cpp not aware of service bits.

theuni commented:
Agree with @TheBlueMatt on taking that as a starting point. We discussed it a little, and I think defining functions for the filtered use-cases makes life much easier. @jonasschnelli it's kinda intrusive on the PR, but would it be possible to work that in?

As for making net not aware of service bits, I'm working on breaking addrman out right now. Should be pretty painless and a nice improvement.

promag commented on test/functional/test_framework/mininode.py:
```diff
@@ -221,19 +221,24 @@ def ToHex(obj):
 
-    def deserialize(self, f):
+    def deserialize(self, f, with_time=True):
+        if with_time == True:
```
Remove `== True`.

promag commented on test/functional/test_framework/mininode.py:
```diff
@@ -221,19 +221,24 @@ def ToHex(obj):
-    def serialize(self):
+    def serialize(self, with_time=True):
         r = b""
+        if with_time == True:
```
Remove `== True`.

promag commented on src/net.cpp:
```diff
@@ -183,6 +183,10 @@ void AdvertiseLocal(CNode *pnode)
     if (fListen && pnode->fSuccessfullyConnected)
     {
         CAddress addrLocal = GetLocalAddress(&pnode->addr, pnode->GetLocalServices());
+        if (gArgs.GetBoolArg("-addrmantest", false) == true) {
```
Remove `== true`.

promag commented on src/net.cpp:
```diff
@@ -191,7 +195,7 @@ void AdvertiseLocal(CNode *pnode)
             addrLocal.SetIP(pnode->GetAddrLocal());
         }
-        if (addrLocal.IsRoutable())
+        if (addrLocal.IsRoutable() || gArgs.GetBoolArg("-addrmantest", false) == true)
```
Remove `== true`.

promag commented on test/functional/node_network_limited.py:
```diff
@@ -0,0 +1,107 @@
+        getdata_request = msg_getdata()
+        getdata_request.inv.append(CInv(2, int(blockhash, 16)))
+        node.send_message(getdata_request)
+        #check if we the peer sends us the block
```
Nit, space after `#` here and below.

promag commented on test/functional/node_network_limited.py:
```diff
@@ -0,0 +1,107 @@
+
+        #NODE_NETWORK_LIMITED must still be signaled after restart
+        self.stop_node(0)
+        self.start_node(0, extra_args=self.extra_args[0])
```
`self.restart_node(0)`.

jonasschnelli committed:
Accept addresses with NODE_NETWORK_LIMITED flag

jonasschnelli committed:
Connect to peers signaling NODE_NETWORK_LIMITED when out-of-IBD

jonasschnelli commented:
Rebased (since #11456 is now merged). Tried to keep calls to `InitialBlockDownload()` to a minimum, therefore added `initialBlockDownloadCompleted` (bool) to `protocol.h`. Not sure if this is the best idea (ping @TheBlueMatt & @theuni).

fanquake commented:
This needs a rebase.

TheBlueMatt commented on src/protocol.h:
```diff
@@ -297,6 +299,9 @@ enum ServiceFlags : uint64_t {
  */
 static ServiceFlags GetDesirableServiceFlags(ServiceFlags services) {
+    if (initialBlockDownloadCompleted) {
+        return ServiceFlags(NODE_NETWORK_LIMITED | NODE_WITNESS);
```
This would result in us *only* connecting to NODE_NETWORK_LIMITED after IBD - I think this should be:
```
if (services & NODE_NETWORK_LIMITED && initialBlockDownloadCompleted) return ServiceFlags(NODE_NETWORK_LIMITED | NODE_WITNESS);
return ServiceFlags(NODE_NETWORK | NODE_WITNESS);
```

TheBlueMatt commented on src/protocol.h:
```diff
@@ -311,10 +311,10 @@ static inline bool HasAllDesirableServiceFlags(ServiceFlags services) {
  */
 static inline bool MayHaveUsefulAddressDB(ServiceFlags services) {
-    return services & NODE_NETWORK;
+    return (services & NODE_NETWORK) || (services & NODE_NETWORK_LIMITED);
```
The commit message is weird here - this function is intended to be used to decide whether a node may be a useful feeler, and thus the commit message should be about relayed addresses *from* nodes with NODE_NETWORK_LIMITED set. But since we also use it in ADDR message processing, it kinda means both - I think we should change the ADDR message processing to MayHaveUsefulAddressDB || HasAllDesirableServiceFlags to clarify things (which will not actually be a chahnge, but thats fine).

jonasschnelli replied:
Fixed

TheBlueMatt reviewed (commented):
Looks pretty good. Didn't review tests, needs rebase as noted.

NicolasDorier commented:
>  NODE_NETWORK_LIMITED supporting peers SHOULD avoid leaking the prune depth and therefore not serve blocks deeper then the signaled NODE_NETWORK_LIMITED thresholds.

could we relax this in case the peer is whitelisted ?

Use case:

I plan to use BIP159 for making it cheaper to host BTCPay. However, it might happen that my whitelisted block explorer crashes more than 2 days for reasons.
I would need then to restart IBD from beginning.

I would like to be able to configure prune to keep like 7 days of blocks, so that my block explorer could come back online after 3 days without issues.

gmaxwell commented:
Perhaps this should be split into "set the limited flag" PR and make use of it, the setting is ultra trivial and could even be considered for backport, esp since the anti-fingerprinting part is a bugfix.

The actual use of the flag in our connection I think needs work,  we should be using our current time and our current headers to decide if LIMITED peers are likely useful to us. IsInitialBlockDownload is a very vague and general criteria and isn't the one we want here.

I think what we want is a function that tells us if we expect limited peers to be useful.   If the headers we have say that there are more than 280 blocks missing,  limited is not useful for us. Or if our best header timestamp is more than (say) 1.75 days old, we should assume they're not useful to us.    When we connect, we'll learn headers, so if for example we're 300 blocks behind we'll find out then, and then not consider limited peers anymore until we catch up.

For deciding on fetching, we should already have headers, and so the fetch should perhaps just only be checking if the peer block is more than 280 blocks back (again, giving 8 blocks slack to cover cases where the peer gets blocks where we haven't heard of the headers yet).

gmaxwell commented on src/net_processing.cpp:
```diff
@@ -1059,6 +1059,13 @@ void static ProcessGetData(CNode* pfrom, const Consensus::Params& consensusParam
+                        (((pfrom->GetLocalServices() & NODE_NETWORK_LIMITED) == NODE_NETWORK_LIMITED) && ((pfrom->GetLocalServices() & NODE_NETWORK) != NODE_NETWORK) && (chainActive.Tip()->nHeight - mi->second->nHeight > (int)NODE_NETWORK_LIMITED_MIN_BLOCKS) )
+                   )) {
+                    LogPrint(BCLog::NET, "Ignore block request below NODE_NETWORK_LIMITED threshold from peer=%d\n", pfrom->GetId());
+                    send = false;
```
This needs to disconnect, I think.  Otherwise we DOS attack our peer, causing them to take a 20 minute timeout on their attempted request to fetch a block they think we're doing to provide.

jonasschnelli commented:
Rebased and worked in @TheBlueMatt and @gmaxwell point.
It's still based on `IsInitialBlockDownload()` but I have plans to implement @gmaxwell idea (which makes sense to me) to rely on out time and headers.

Sjors commented:
This 288 blocks limit seems a bit overzealous, but I suppose it can be expanded later if nodes just assume 288 for un-upgraded peers.

I couldn't find the discussion that led to that [change in BIP159](https://github.com/bitcoin/bips/commit/ee45bbc7c4962d4b551d730469749bb35df18757#diff-6535b2c529c4fce0dc36d586cc0b05f4), so maybe this was already brought up.

The node on my laptop is pruned to 10 GB, Bitseed and Raspberry Pie nodes might be pruned to 100-200 GB. They can be of use to nodes even during IBD. To prevent fingerprinting and still allow more flexibility in sharing data, perhaps `NODE_NETWORK_LIMITED_MIN_BLOCKS` can be made configurable and somehow communicated? The default could still be 288, with a number of specific suggestions in bitcoin.conf.

The setting could be sanity checked against `prune=` and - better, but not sure if possible - automatically chosen based on the worst case for `prune=`.

Also, what happens if a node has a prune= setting that's above the current blockchain size? Does it advertise as NODE_NETWORK_LIMITED? Ideally it wouldn't. Perhaps this can be added to the BIP?

Sjors commented:
Found this (Google is more helpful when you search by name than by BIP number): https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2017-May/014314.html (should be added to the References section in the BIP?)

From that discussion:
> The peak at syncing is at ~24 hours

This explains why even 288 is very useful (worth mentioning in the BIP).

There was some discussion around which sizes to allow, but it was mostly focussed on not using more service bits than needed (or states). I wouldn't want to waste those of course. Can a node simply disclose the pruning depth when asked (within a limited set of choices to prevent fingerprinting), or would that open cans of worms?

jonasschnelli commented:
I think we should only discuss the implementation here. The BIP discussion belongs to the mailing list (https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2017-May/thread.html#14315).

Revealing the prune depth is considered a fingerprinting weakness.

The current BIP and implementation is extendable. I'd like to start with a sane buffer 24h of blocks (144) + 144 block reorganisation buffer. It is possible to extend this later with more differentiated depths.

It would also be possible to bypass the anti-fingerprinting hard 288blocks limit when whitelisting the peer (or once we have BIP150) (not in scope for this PR).

promag commented on doc/bips.md:
```diff
@@ -33,3 +33,4 @@ BIPs that are implemented by Bitcoin Core (up-to-date up to **v0.13.0**):
 * [`BIP 145`](https://github.com/bitcoin/bips/blob/master/bip-0145.mediawiki): getblocktemplate updates for Segregated Witness as of **v0.13.0** ([PR 8149](https://github.com/bitcoin/bitcoin/pull/8149)).
 * [`BIP 147`](https://github.com/bitcoin/bips/blob/master/bip-0147.mediawiki): NULLDUMMY softfork as of **v0.13.1** ([PR 8636](https://github.com/bitcoin/bitcoin/pull/8636) and [PR 8937](https://github.com/bitcoin/bitcoin/pull/8937)).
 * [`BIP 152`](https://github.com/bitcoin/bips/blob/master/bip-0152.mediawiki): Compact block transfer and related optimizations are used as of **v0.13.0** ([PR 8068](https://github.com/bitcoin/bitcoin/pull/8068)).
+* [`BIP 159`](https://github.com/bitcoin/bips/blob/master/bip-0159.mediawiki): NODE_NETWORK_LIMITED service bits are supported as of **v0.16.0** ([PR 10387](https://github.com/bitcoin/bitcoin/pull/10387)).
```
service bit (singular)?

promag commented on test/functional/node_network_limited.py:
```diff
@@ -0,0 +1,100 @@
+
+    def run_test(self):
+        #NODE_BLOOM & NODE_WITNESS & NODE_NETWORK_LIMITED must now be signaled
+        assert_equal(self.getSignaledServiceFlags(), 1036) #1036 == 0x40C == 0100 0000 1100
```
Add space after `#` so that the arrows below point to the correct bits.

promag reviewed (commented):
Needs rebase.

jonasschnelli committed:
[QA] fix mininode CAddress ser/deser

TheBlueMatt commented on src/net_processing.cpp:
```diff
@@ -938,11 +938,12 @@ void PeerLogicValidation::BlockChecked(const CBlock& block, const CValidationSta
-        if (it != mapBlockSource.end()) {
-            MaybeSetPeerAsAnnouncingHeaderAndIDs(it->second.first, connman);
+    else if (state.IsValid()) {
+        initialBlockDownloadCompleted = !IsInitialBlockDownload();
```
Can you move this update to UpdatedBlockTip? (it already has an fInitialDownload instead of having to call IsInitialDownload, even).

jonasschnelli replied:
This code part doesn't add a new call, it just preserves the state of the result of `IsInitialBlockDownload()` in the new `IsInitialBlockDownload` variable.

TheBlueMatt replied:
Yes, I'm suggesting you move the update of initialBlockDownloadCompleted to UpdatedBlockTip instead of BlockChecked.

TheBlueMatt reviewed (commented):
I think IsInitialBlockDownload is fine for now. It requires that the block time on our block tip be at least within the last day, which is a reasonable metric for if LIMITED peers are useful. Sadly we don't have a concept of our best header against which to test, and I'm not sure we want to decide LIMITED peers are not useful if we find a chain (eg fork that pays out 10x reward to encourage miners temporarily) that is invalid.

jonasschnelli commented:
Rebased and removed the now merged signalling part.
This PR adds address relay and eventually connecting to NODE_NETWORK_LIMITED if we are out of IBD.
Not sure, but maybe @gmaxwell [idea ](https://github.com/bitcoin/bitcoin/pull/10387#issuecomment-343357330) should be implemented later?

NicolasDorier commented on src/protocol.h:
```diff
@@ -294,6 +296,7 @@ enum ServiceFlags : uint64_t {
  * Thus, generally, avoid calling with peerServices == NODE_NONE.
  */
 static ServiceFlags GetDesirableServiceFlags(ServiceFlags services) {
+    if (services & NODE_NETWORK_LIMITED && initialBlockDownloadCompleted) return ServiceFlags(NODE_NETWORK_LIMITED | NODE_WITNESS);
```
On https://github.com/bitcoin/bitcoin/pull/10387/files#diff-eff7adeaec73a769788bb78858815c91R1625 you are defining a limitedPeer as one which does not have `NODE_NETWORK` AND which have `NODE_NETWORK_LIMITED`. Which is different from this logic.

jonasschnelli replied:
I can't follow you comment.
`pfrom->fLimitedNode` is set to true if the peer does not signal `NODE_NETWORK` but signals `NODE_NETWORK_LIMITED`.
At this point `GetDesirableServiceFlags()`, we should return `NODE_NETWORK_LIMITED` (not `NODE_NETWORK`) if we are out of IBD. I think this is correct,...

NicolasDorier replied:
In other words, this is what I would have done:

```
var limited =  (!(services & NODE_NETWORK) && (services & NODE_NETWORK_LIMITED));
if (limited && initialBlockDownloadCompleted) return ServiceFlags(NODE_NETWORK_LIMITED | NODE_WITNESS);
return ServiceFlags(NODE_NETWORK | NODE_WITNESS);
```

The reason is that this is how you define if a peer is limited from his services flags [before]( https://github.com/bitcoin/bitcoin/pull/10387/files#diff-eff7adeaec73a769788bb78858815c91R1625), I don't understand why you are using different logic here.

jonasschnelli replied:
I don't think we should check for `!(services & NODE_NETWORK)` because we also want to connect to NODE_NETWORK peers... but we are also happy by connecting to a peer that has `NODE_NETWORK_LIMITED` but not `NODE_NETWORK`.

NicolasDorier replied:
Imagine the following scenario:

* You are not in IBD, connect to peers only with `NODE_NETWORK|NODE_NETWORK_LIMITED`
* For some reason, you start being in IBD once again.

Now your node is stuck because of [this](https://github.com/bitcoin/bitcoin/pull/10387/files#diff-eff7adeaec73a769788bb78858815c91R3598).

An alternative fix would be to only check for `NODE_NETWORK_LIMITED` [here](https://github.com/bitcoin/bitcoin/pull/10387/files#diff-eff7adeaec73a769788bb78858815c91R1625).

jonasschnelli replied:
> In general I think it is weird that we treat differently NODE_NETWORK|NODE_NETWORK_LIMITED peer from NODE_NETWORK peers.

I think we have to. Otherwise we would never connect to pruned peers... what we want is also considering connecting to pruned if we are not deeper then 144 blocks from the expected tip.

> For some reason, you start being in IBD once again.

I think there you are right. When switching back to IBD below say 144*1.5 blocks from the tip, we eventually hit a peer stalling us. Though it should be detected by the stalling logic and should get disconnected. The free slow should be then be re-used with a peer that must have `NODE_NETWORK` (because `GetDesirableServiceFlags()` will required `NODE_NETWORK` again at this point).

But I agree that disconnecting pruned (`NODE_NETWORK_LIMITED` only) peers once we fallback to IBD should be something we implement.

NicolasDorier replied:
`NODE_NETWORK|NODE_NETWORK_LIMITED` makes no sense to me. I was under the impression that if a peer was serving only limited block it would be `NODE_NETWORK_LIMITED`. If the peer can serve more, it would be `NODE_NETWORK`.

So what is the meaning of a peer `NODE_NETWORK|NODE_NETWORK_LIMITED` ? I would argue that this mean they can serve historical blocks, so they should be considered as if they were `NODE_NETWORK` only.

jonasschnelli replied:
As of BIP159, `NODE_NETWORK_LIMITED` signals that a peer has the last 288 blocks. This is "always" true for a NODE_NETWORK peer. This was chosen because services are a bit-set and bits should not depend on each other.

Conclusion:
* `NODE_NETWORK` (unpruned peer, can serve the complete history)
* `NODE_NETWORK_LIMITED ^NODE_NETWORK` (no NODE_NETWORK) (can serve the last 288 blocks but can't serve the complete history)
* `NODE_NETWORK | NODE_NETWORK_LIMITED` (can serve the complete history and the last 288 blocks [which is obvious]

NicolasDorier replied:
OK I got it now. Thanks for explanation, this is fine to me.

NicolasDorier commented:
Code review ACK. Might need in a follow up PR to kick limited peers when we go back in IBD though.

jimpo commented on test/functional/test_framework/messages.py:
```diff
@@ -849,11 +854,11 @@ def deserialize(self, f):
         self.nTime = struct.unpack("<q", f.read(8))[0]
         self.addrTo = CAddress()
-        self.addrTo.deserialize(f)
+        self.addrTo.deserialize(f, False)
```
I think it's more clear to call with keyword arguments for boolean options: `deserialize(f, with_time=False)`. Same for invocations below.

laanwj replied:
Agree; using a named argument would make this code more readable.

jimpo reviewed (commented):
A few stylistic comments.

jonasschnelli committed:
[QA] add NODE_NETWORK_LIMITED address relay and sync test

jonasschnelli commented:
Rebased (needed due to the the node network test overhaul).
Fixed @jimpo's points.

jonasschnelli commented:
Rebased.

laanwj commented:
utACK 3f56df5

laanwj commented on src/protocol.h:
```diff
@@ -277,6 +278,7 @@ enum ServiceFlags : uint64_t {
     // BIP process.
 };
 
+extern std::atomic<bool> g_initial_block_download_completed;
```
I don't like that we need to introduce a global here, especially one that has to be accessible externally (in the header file).
Any drawback to moving `GetDesirableServiceFlags` to the .cpp instead?

jonasschnelli committed:
Add setter for g_initial_block_download_completed

jonasschnelli commented:
Added a commit that hides the new `protocol.h/.cpp` global (`g_initial_block_download_completed`) behinds a setter (@laanwj / https://github.com/bitcoin/bitcoin/pull/10387#discussion_r168228007).

kaycoinsUSA commented:
updating ,great work.
